File: library/utils/src/main/java/com/example/utils/storage/StorageManagerUtil.java
package com.example.utils.storage;

import android.content.Context;
import android.text.TextUtils;
import android.util.LruCache;

import java.util.HashMap;
import java.util.Map;

/**
 * 应用存储管理，不建议使用SharedPreferencesUtil存储数据
 */
public class StorageManagerUtil {

    private static final String KEY_DEFAULT = "KEY_DEFAULT";

    private static Map<String, LruCache<String, Object>> lruCacheMap;

    private static int sMaxSize = 20;

    public static void init(Context context, int maxSize) {
        SharedPreferencesUtil.init(context);
        lruCacheMap = new HashMap<>();
        sMaxSize = maxSize;
    }

    public static void init(Context context) {
        SharedPreferencesUtil.init(context);
        lruCacheMap = new HashMap<>();
    }

    /**
     * 清除缓存
     */
    public static void clearCache() {
        for (LruCache<String, Object> objectLruCache : lruCacheMap.values()) {
            objectLruCache.evictAll();
        }
    }

    /**
     * 清除指定模块的缓存
     */
    public static void clearCache(String module) {
        if (lruCacheMap.containsKey(module) && null != lruCacheMap.get(module)) {
            lruCacheMap.get(module).evictAll();
        }
    }

    /**
     * 清除指定模块下指定key的缓存
     */
    public static void clearCache(String module, String key) {
        if (lruCacheMap.containsKey(module) && null != lruCacheMap.get(module)) {
            lruCacheMap.get(module).remove(key);
        }
    }

    /**
     * 清除指定模块的本地数据
     */
    public static void clearData(String module) {
        SharedPreferencesUtil.getInstance(module).clear();
    }

    /**
     * 清除指定模块下指定key的本地数据
     */
    public static void clearData(String module, String key) {
        SharedPreferencesUtil.getInstance(module).remove(key);
    }

    /**
     * 存储到本地
     */
    public static void putToDisk(String key, String value) {
        putToDisk(KEY_DEFAULT, key, value);
    }

    public static void putToDisk(String module, String key, String value) {
        putToCache(module, key, value);
        SharedPreferencesUtil.getInstance(module).put(key, value);
    }

    public static void putToDisk(String key, int value) {
        putToDisk(KEY_DEFAULT, key, value);
    }

    public static void putToDisk(String module, String key, int value) {
        putToCache(module, key, value);
        SharedPreferencesUtil.getInstance(module).put(key, value);
    }

    public static void putToDisk(String key, long value) {
        putToDisk(KEY_DEFAULT, key, value);
    }

    public static void putToDisk(String module, String key, long value) {
        putToCache(module, key, value);
        SharedPreferencesUtil.getInstance(module).put(key, value);
    }

    public static void putToDisk(String key, boolean value) {
        putToDisk(KEY_DEFAULT, key, value);
    }

    public static void putToDisk(String module, String key, boolean value) {
        putToCache(module, key, value);
        SharedPreferencesUtil.getInstance(module).put(key, value);
    }

    public static void putToDisk(String key, float value) {
        putToDisk(KEY_DEFAULT, key, value);
    }

    public static void putToDisk(String module, String key, float value) {
        putToCache(module, key, value);
        SharedPreferencesUtil.getInstance(module).put(key, value);
    }

    /**
     * 获取数据
     */
    public static <T> T get(String key, T defaultValue) {
        return get(KEY_DEFAULT, key, defaultValue);
    }

    public static <T> T get(String module, String key, T defaultValue) {
        if (TextUtils.isEmpty(module)) {
            return defaultValue;
        }
        if (TextUtils.isEmpty(key)) {
            return defaultValue;
        }
        Object v = getFromCache(module, key);
        if (null == v) {
            return defaultValue;
        }
        return (T) v;
    }

    public static String getString(String key) {
        return get(key, "");
    }

    public static String get(String key, String defaultValue) {
        return get(KEY_DEFAULT, key, defaultValue);
    }

    public static String get(String module, String key, String defaultValue) {
        if (TextUtils.isEmpty(module)) {
            return defaultValue;
        }
        if (TextUtils.isEmpty(key)) {
            return defaultValue;
        }
        Object v = getFromCache(module, key);
        if (null == v) {
            if (null == defaultValue) {
                defaultValue = "";
            }
            String s = SharedPreferencesUtil.getInstance(module).get(key, defaultValue);
            putToCache(module, key, s);
            return s;
        }
        if (v instanceof String) {
            return (String) v;
        }
        return defaultValue;
    }

    public static int getInt(String key) {
        return get(key, 0);
    }

    public static int get(String key, int defaultValue) {
        return get(KEY_DEFAULT, key, defaultValue);
    }

    public static int get(String module, String key, int defaultValue) {
        if (TextUtils.isEmpty(module)) {
            return defaultValue;
        }
        if (TextUtils.isEmpty(key)) {
            return defaultValue;
        }
        Object v = getFromCache(module, key);
        if (null == v) {
            int i = SharedPreferencesUtil.getInstance(module).get(key, defaultValue);
            putToCache(module, key, i);
            return i;
        }
        if (v instanceof Integer) {
            return (int) v;
        }
        return defaultValue;
    }

    public static long getLong(String key) {
        return get(key, 0L);
    }

    public static long get(String key, long defaultValue) {
        return get(KEY_DEFAULT, key, defaultValue);
    }

    public static long get(String module, String key, long defaultValue) {
        if (TextUtils.isEmpty(module)) {
            return defaultValue;
        }
        if (TextUtils.isEmpty(key)) {
            return defaultValue;
        }
        Object v = getFromCache(module, key);
        if (null == v) {
            long l = SharedPreferencesUtil.getInstance(module).get(key, defaultValue);
            putToCache(module, key, l);
            return l;
        }
        if (v instanceof Long) {
            return (long) v;
        }
        return defaultValue;
    }

    public static boolean getBoolean(String key) {
        return get(key, false);
    }

    public static boolean get(String key, boolean defaultValue) {
        return get(KEY_DEFAULT, key, defaultValue);
    }

    public static boolean get(String module, String key, boolean defaultValue) {
        if (TextUtils.isEmpty(module)) {
            return defaultValue;
        }
        if (TextUtils.isEmpty(key)) {
            return defaultValue;
        }
        Object v = getFromCache(module, key);
        if (null == v) {
            boolean b = SharedPreferencesUtil.getInstance(module).get(key, defaultValue);
            putToCache(module, key, b);
            return b;
        }
        if (v instanceof Boolean) {
            return (boolean) v;
        }
        return defaultValue;
    }

    public static float getFloat(String key) {
        return get(key, 0f);
    }

    public static float get(String key, float defaultValue) {
        return get(KEY_DEFAULT, key, defaultValue);
    }

    public static float get(String module, String key, float defaultValue) {
        if (TextUtils.isEmpty(module)) {
            return defaultValue;
        }
        if (TextUtils.isEmpty(key)) {
            return defaultValue;
        }
        Object v = getFromCache(module, key);
        if (null == v) {
            Float f = SharedPreferencesUtil.getInstance(module).get(key, defaultValue);
            putToCache(module, key, f);
            return f;
        }
        if (v instanceof Float) {
            return (float) v;
        }
        return defaultValue;
    }

    /**
     * 存储到缓存
     */
    private static void putToCache(String module, String key, Object value) {
        try {
            if (!lruCacheMap.containsKey(module) && null == lruCacheMap.get(module)) {
                lruCacheMap.put(module, new LruCache<String, Object>(sMaxSize));
            }
            if (null != value && null != lruCacheMap.get(module)) {
                lruCacheMap.get(module).put(key, value);
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    private static Object getFromCache(String module, String key) {
        if (null == lruCacheMap || !lruCacheMap.containsKey(module) || null == lruCacheMap.get(module)) {
            return null;
        }
        return lruCacheMap.get(module).get(key);
    }
}

File: library/utils/src/main/java/com/example/utils/storage/StorageManagerUtil.kt
package com.example.utils.storage

import android.content.Context
import android.util.LruCache
import com.example.utils.storage.SharedPreferencesUtil.Companion.getInstance
import java.util.*

/**
 * 应用存储管理，不建议使用SharedPreferencesUtil存储数据
 */
object StorageManagerUtil {

    private const val KEY_DEFAULT = "KEY_DEFAULT"
    private var lruCacheMap = HashMap<String, LruCache<String, Any?>>()
    private var sMaxSize = 20

    fun init(context: Context, maxSize: Int) {
        SharedPreferencesUtil.init(context)
        sMaxSize = maxSize
    }

    fun init(context: Context) {
        SharedPreferencesUtil.init(context)
    }

    /**
     * 清除缓存
     */
    fun clearCache() {
        for (objectLruCache in lruCacheMap.values) {
            objectLruCache.evictAll()
        }
    }

    /**
     * 清除指定模块的缓存
     */
    fun clearCache(module: String?) {
        lruCacheMap[module]?.evictAll()
    }

    /**
     * 清除指定模块下指定key的缓存
     */
    fun clearCache(module: String?, key: String?) {
        lruCacheMap[module]?.remove(key)
    }

    /**
     * 清除指定模块的本地数据
     */
    fun clearData(module: String) {
        getInstance(module).clear()
    }

    /**
     * 清除指定模块下指定key的本地数据
     */
    fun clearData(module: String, key: String) {
        getInstance(module).remove(key)
    }

    /**
     * 存储到本地
     */
    fun putToDisk(key: String, value: String?) {
        putToDisk(KEY_DEFAULT, key, value)
    }

    @JvmStatic
    fun putToDisk(module: String, key: String, value: String?) {
        putToCache(module, key, value)
        getInstance(module).put(key, value)
    }

    fun putToDisk(key: String, value: Int) {
        putToDisk(KEY_DEFAULT, key, value)
    }

    @JvmStatic
    fun putToDisk(module: String, key: String, value: Int) {
        putToCache(module, key, value)
        getInstance(module).put(key, value)
    }

    fun putToDisk(key: String, value: Long) {
        putToDisk(KEY_DEFAULT, key, value)
    }

    @JvmStatic
    fun putToDisk(module: String, key: String, value: Long) {
        putToCache(module, key, value)
        getInstance(module).put(key, value)
    }

    fun putToDisk(key: String, value: Boolean) {
        putToDisk(KEY_DEFAULT, key, value)
    }

    @JvmStatic
    fun putToDisk(module: String, key: String, value: Boolean) {
        putToCache(module, key, value)
        getInstance(module).put(key, value)
    }

    fun putToDisk(key: String, value: Float) {
        putToDisk(KEY_DEFAULT, key, value)
    }

    @JvmStatic
    fun putToDisk(module: String, key: String, value: Float) {
        putToCache(module, key, value)
        getInstance(module).put(key, value)
    }

    /**
     * 获取数据
     */
    @JvmStatic
    operator fun <T> get(key: String, defaultValue: T): T {
        return get(KEY_DEFAULT, key, defaultValue)
    }

    @JvmStatic
    operator fun <T> get(module: String, key: String, defaultValue: T): T {
        val v = getFromCache(module, key) ?: return defaultValue
        return v as T
    }

    fun getString(key: String): String {
        return get(key, "")
    }

    @JvmStatic
    operator fun get(key: String, defaultValue: String?): String {
        return get(KEY_DEFAULT, key, defaultValue)
    }

    @JvmStatic
    operator fun get(
        module: String,
        key: String,
        defaultValue: String?
    ): String {
        val v = getFromCache(module, key)
        if (null == v) {
            val s = getInstance(module)[key, defaultValue] ?: ""
            putToCache(module, key, s)
            return s
        }
        return if (v is String) {
            v
        } else defaultValue ?: ""
    }

    fun getInt(key: String): Int {
        return get(key, 0)
    }

    @JvmStatic
    operator fun get(key: String, defaultValue: Int): Int {
        return get(KEY_DEFAULT, key, defaultValue)
    }

    @JvmStatic
    operator fun get(module: String, key: String, defaultValue: Int): Int {
        val v = getFromCache(module, key)
        if (null == v) {
            val i = getInstance(module)[key, defaultValue]
            putToCache(module, key, i)
            return i
        }
        return if (v is Int) {
            v
        } else defaultValue
    }

    fun getLong(key: String): Long {
        return get(key, 0L)
    }

    @JvmStatic
    operator fun get(key: String, defaultValue: Long): Long {
        return get(KEY_DEFAULT, key, defaultValue)
    }

    @JvmStatic
    operator fun get(module: String, key: String, defaultValue: Long): Long {
        val v = getFromCache(module, key)
        if (null == v) {
            val l = getInstance(module)[key, defaultValue]
            putToCache(module, key, l)
            return l
        }
        return if (v is Long) {
            v
        } else defaultValue
    }

    fun getBoolean(key: String): Boolean {
        return get(key, false)
    }

    @JvmStatic
    operator fun get(key: String, defaultValue: Boolean): Boolean {
        return get(KEY_DEFAULT, key, defaultValue)
    }

    @JvmStatic
    operator fun get(
        module: String,
        key: String,
        defaultValue: Boolean
    ): Boolean {
        val v = getFromCache(module, key)
        if (null == v) {
            val b = getInstance(module)[key, defaultValue]
            putToCache(module, key, b)
            return b
        }
        return if (v is Boolean) {
            v
        } else defaultValue
    }

    fun getFloat(key: String): Float {
        return get(key, 0f)
    }

    @JvmStatic
    operator fun get(key: String, defaultValue: Float): Float {
        return get(KEY_DEFAULT, key, defaultValue)
    }

    @JvmStatic
    operator fun get(module: String, key: String, defaultValue: Float): Float {
        val v = getFromCache(module, key)
        if (null == v) {
            val f = getInstance(module)[key, defaultValue]
            putToCache(module, key, f)
            return f
        }
        return if (v is Float) {
            v
        } else defaultValue
    }

    /**
     * 存储到缓存
     */
    private fun putToCache(
        module: String,
        key: String,
        value: Any?
    ) {
        try {
            if (null == lruCacheMap[module]) {
                lruCacheMap[module] = LruCache(sMaxSize)
            }
            lruCacheMap[module]!!.put(key, value)
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    private fun getFromCache(module: String, key: String): Any? {
        return lruCacheMap[module]?.get(key)
    }
}